src/chainstate/stacks/index/proofs.rs

```diff
@@ -0,0 +1,1135 @@
+                    Some(bytes) => {
+                        bytes
+                    },
+                    None => {
+                        trace!("Failed to get the path from the proof");
+                        return false;
+                    }
+                };
+            }
+            else {
+                // make sure that this segment proof is a prefix of the last
+                let new_path_bytes = match TrieMerkleProof::get_segment_proof_path_prefix(segment_proof) {
+                    Some(bytes) => {
+                        bytes
+                    },
+                    None => {
+                        trace!("Failed to et the path prefix from the proof");
+                        return false;
+                    }
+                };
+
+                if path_bytes.len() < new_path_bytes.len() {
+                    trace!("Segment proof path is {}, which is longer than the previous segment proof length {}", path_bytes.len(), new_path_bytes.len());
+                    trace!("path_bytes: {:?}", &path_bytes);
+                    trace!("new path bytes: {:?}", &new_path_bytes);
+                    return false;
+                }
+
+                for i in 0..new_path_bytes.len() {
+                    if path_bytes[i] != new_path_bytes[i] {
+                        trace!("Segment path {:?} is not a prefix of previous segment path {:?}", &new_path_bytes, &path_bytes);
+                        return false;
+                    }
+                }
+
+                path_bytes = new_path_bytes;
+            }
+
+            // next shunt proof 
+            i = j;
+            if i >= proof.len() {
+                trace!("Proof is incomplete -- must end with a shunt proof");
+                return false;
+            }
+
+            j = i + 1;
+            while j < proof.len() {
+                match proof[j] {
+                    TrieMerkleProofType::Shunt(_) => {
+                        j += 1;
+                    },
+                    _ => {
+                        break;
+                    }
+                }
+            }
+
+            // end of shunt proof
+            i = j;
+        }
+
+        true
+    }
+
+
+    /// Given a value and the root hash from which this proof was
+    /// (supposedly) generated go and verify whether or not it is consistent with the root hash.
+    /// For the proof validation to work, the verifier needs to know which Trie roots correspond to
+    /// which block headers.  This can be calculated and verified independently from the blockchain
+    /// headers.
+    /// NOTE: Trie root hashes are globally unique by design, even if they represent the same contents, so the root_to_block map is bijective with high probability.
+    pub fn verify_proof(proof: &Vec<TrieMerkleProofType>, root_hash: &TrieHash, root_to_block: &HashMap<TrieHash, BlockHeaderHash>) -> bool {
```

Comment:
Feedback/question, regarding the MARF Merkle proof.

At this point, the proof verification process helps asserting that “a given key exists at this particular block”, but does not help checking the value of the key. 
The proof constructor does this job here: 
https://github.com/blockstack/blockstack-core/blob/f2322c2fffb8c0b01c0ecfa07192870facb7a743/src/chainstate/stacks/index/proofs.rs#L1088-L1091
However Merkle proofs are meant to be manipulated off-chain (at least in my mind 🙂), and no one can have the guaranty that a given proof have been build with this function.
This nuance is surely irrelevant in the cases of transactions, since we will probably use the transaction hash as a key. Knowing that “a given key exists at this particular block” will naturally be implying that the transaction is part of this block.

However, when something else than the value’s hash is used as a key, the construct might look incomplete, because the proof just informs you that a given key exists, without telling you much about its value. 
That could maybe be fine in a world where double insertions would be rejected but since we’re updating the values, isn’t it an issue?

```rust
    #[test]
    fn toying_with_proofs() {
        let path = "/tmp/toying_with_marf".to_string();
        match fs::metadata(&path) {
            Ok(_) => {
                fs::remove_dir_all(&path).unwrap();
            },
            Err(_) => {}
        };

        let mut m = MARF::from_path(&path).unwrap();

        let sentinel_block = TrieFileStorage::block_sentinel();
        let block_0 = BlockHeaderHash([0u8; 32]);
        let block_1 = BlockHeaderHash([1u8; 32]);
        let block_2 = BlockHeaderHash([2u8; 32]);

        let k1 = "K1".to_string();
        let old_v = "OLD".to_string();
        let new_v = "NEW".to_string();

        m.begin(&sentinel_block, &block_0).unwrap();
        m.commit().unwrap();

        // Block #1
        m.begin(&block_0, &block_1).unwrap();
        let r = m.insert(&k1, &old_v);
        m.commit();

        // Block #2
        m.begin(&block_1, &block_2).unwrap();
        let r = m.insert(&k1, &new_v);
        m.commit();

        let old_value = m.get(&block_1, &k1);
        println!("OLD: {:?}", old_value);

        let new_value = m.get(&block_2, &k1).unwrap().unwrap();
        println!("NEW: {:?}", new_value);
        
        let path = TriePath::from_key(&k1);

        // Pass :)
        merkle_test_marf_key_value(&mut m.storage, &block_2, &k1, &new_v);

        // Pass :( after commenting proofs.rs:1090
        merkle_test_marf_key_value(&mut m.storage, &block_2, &k1, &old_v);
    }
```

Reply:
Looks like you found a bug in the verifier :)

If you commented out line 1090, then `merkle_test_marf_key_value` will generate a proof that `k1` equals `old_v`, which is not the proof the verifier asked for.  A correct client needs to be able to determine this; the given implementation does not do this.

The bug is that the verifier should take the key and value it requested the proof for, and _additionally_ verify two more things:

* The hash of the value must be equal to the first segment proof hash.  This will (correctly) cause the verifier to fail in this example, since `hash(new_v)` is not equal to `hash(old_v)`.
* The verifier needs to check that the set of segment proofs are all prefixes of the hash of the key, and that the value it requested the proof for is at the end of the path.

I will make these fixes and adapt your test-case to make sure they work.

Reply:
Pushed a fix.

Reply:
Nice :)
Do you think we should also update SIP004?

Reply:
Yup; updated.